# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:

import warnings
import numpy as np
import sqlite3
import pandas as pd
warnings.simplefilter('ignore')

C:\Users\arcde\AppData\Local\Temp\ipykernel_20816\476482988.py:4: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [3]:
global con
def dataframe() : 
    try : 
        global con 
        con = sqlite3.connect("Assets/Week 2/go_sales.sqlite")
        print("Connected to sqlite.")
    except sqlite3.Error as error: 
        print("Failed to read data from sqlite table", error)
    

dataframe()
print(con)

Connected to sqlite.


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [4]:
global product
product = pd.read_sql_query("SELECT * FROM {}".format("product"),con)

global product_type
product_type = pd.read_sql_query("SELECT * FROM {}".format("product_type"),con)

global product_line
product_line = pd.read_sql_query("SELECT * FROM {}".format("product_line"),con)

global sales_staff
sales_staff = pd.read_sql_query("SELECT * FROM {}".format("sales_staff"),con)

global sales_branch
sales_branch = pd.read_sql_query("SELECT * FROM {}".format("sales_branch"),con)

global retailer_site
retailer_site = pd.read_sql_query("SELECT * FROM {}".format("retailer_site"),con)

global country
country = pd.read_sql_query("SELECT * FROM {}".format("country"),con)

global order_header
order_header = pd.read_sql_query("SELECT * FROM {}".format("order_header"),con)

global order_details
order_details = pd.read_sql_query("SELECT * FROM {}".format("order_details"),con)

global target
target = pd.read_sql_query("SELECT * FROM {}".format("sales_targetdata"),con)

global returned_item
returned_item = pd.read_sql_query("SELECT * FROM {}".format("returned_item"),con)

global return_reason
return_reason = pd.read_sql_query("SELECT * FROM {}".format("return_reason"),con)



Columntypes converteren

In [9]:
product["PRODUCTION_COST"] = product["PRODUCTION_COST"].astype(float)
product["MARGIN"] = product["MARGIN"].astype(float)
print(product.dtypes)

PRODUCT_NUMBER        object
INTRODUCTION_DATE     object
PRODUCT_TYPE_CODE     object
PRODUCTION_COST      float64
MARGIN               float64
PRODUCT_IMAGE         object
LANGUAGE              object
PRODUCT_NAME          object
DESCRIPTION           object
TRIAL888              object
dtype: object


Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [47]:
def production_cost_Dataframe() :
    product["PRODUCTION_COST"] = product["PRODUCTION_COST"].astype(float)
    productionconstresult_filter = (product["PRODUCTION_COST"] > 50) & (product["PRODUCTION_COST"] < 100)
    productioncostresult = product.loc[(productionconstresult_filter), ['PRODUCT_NAME','PRODUCTION_COST']]
    print(productioncostresult.count())
    display(productioncostresult)
    # # print(product.keys())
production_cost_Dataframe()

PRODUCT_NAME       23
PRODUCTION_COST    23
dtype: int64


PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [48]:
def product_margin_between() :    
    product_margin_filter = (product["MARGIN"]< 0.20) | (product['MARGIN']>0.60)
    product_margin_result = product.loc[(product_margin_filter),['PRODUCT_NAME','MARGIN']]
    print(product_margin_result.count())
    display(product_margin_result)
product_margin_between()

PRODUCT_NAME    7
MARGIN          7
dtype: int64


PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [49]:
def country_francs() :
    country_currency_filter = (country["CURRENCY_NAME"] == "francs")
    country_currency = country.loc[(country_currency_filter),["COUNTRY"]]
    country_currency_sorted = country_currency.sort_values("COUNTRY")
    print(country_currency_sorted.count())
    display(country_currency_sorted)
country_francs()


COUNTRY    3
dtype: int64


COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [50]:
def introduction_margins() :
    introduction_margin_filter = product["MARGIN"]>0.50
    introduction_margin = product.loc[(introduction_margin_filter),["INTRODUCTION_DATE"]].drop_duplicates("INTRODUCTION_DATE")
    print(introduction_margin.count())
    display(introduction_margin)
introduction_margins()

INTRODUCTION_DATE    7
dtype: int64


INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [51]:
def first_addres() :
    sales_branch_filter = (sales_branch["ADDRESS2"].notna()) & (sales_branch["REGION"].notna())
    sales_branch_location = sales_branch.loc[(sales_branch_filter),["ADDRESS1","CITY"]]
    print(sales_branch_location.count())
    display(sales_branch_location)
first_addres()

ADDRESS1    7
CITY        7
dtype: int64


ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [52]:
def country_dollars() :
    country_dollar_filter = (country["CURRENCY_NAME"].str.contains("dollar"))
    country_dollar_currency = country.loc[(country_dollar_filter),["COUNTRY"]]
    print(country_dollar_currency.count())
    display(country_dollar_currency)
country_dollars()

COUNTRY    4
dtype: int64


COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [60]:
country_postal_filter = ( sales_branch["POSTAL_ZONE"].str[0]=="D")
country_postal = sales_branch.loc[(country_postal_filter),["ADDRESS1","ADDRESS2","CITY"]]
country_postal_sorted = country_postal.sort_values("ADDRESS2")
print(country_postal_sorted.count())
display(country_postal_sorted)

ADDRESS1    2
ADDRESS2    0
CITY        2
dtype: int64


ADDRESS1 ADDRESS2     CITY
0    Schwabentor 35     None  Hamburg
1  Leopoldstraße 36     None  München

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [61]:
print(len(returned_item))

706


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [62]:
print(len(pd.unique(sales_branch["REGION"])))

16


Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [77]:
# print(product.keys())
min = product["MARGIN"].min()
avg = product["MARGIN"].mean()
max = product["MARGIN"].max()


df = pd.DataFrame(
     {
        "Min":min,
        "Avg":avg,
        "Max":max
     }, index=[0]
    )
display(df)


Min       Avg  Max
0  0.17  0.396609  0.7

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [78]:
print(len(sales_branch[sales_branch["ADDRESS2"].isna()]))

19


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [ ]:
def calculate_average_sales_price() : 
    korting_filter = (order_details["UNIT_SALE_PRICE"])

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [5]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [6]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 